network/p2p/connManager.go

```diff
@@ -87,3 +98,47 @@ func (c ConnManager) logConnectionUpdate(n network.Network, con network.Conn, lo
 		Int("total_connections", len(n.Conns())).
 		Msg(logMsg)
 }
+
+// ProtectPeer increments the stream setup count for the peer.ID
+func (c *ConnManager) ProtectPeer(id peer.ID) {
+	c.streamSetupMapLk.Lock()
+	defer c.streamSetupMapLk.Unlock()
+
+	c.streamSetupInProgressCnt[id]++
+
+	c.log.Trace().
+		Str("peer_id", id.String()).
+		Int("stream_setup_in_progress_cnt", c.streamSetupInProgressCnt[id]).
+		Msg("protected from connection pruning")
+}
+
+// UnprotectPeer decrements the stream setup count for the peer.ID.
+// If the count reaches zero, the id is removed from the map
+func (c *ConnManager) UnprotectPeer(id peer.ID) {
```

Comment:
When the stream creation is done (successfully on unsuccessfully) we mark it as unprotected